build-tools/xaprepare/xaprepare/Application/GeneratedSourceLinkJsonFile.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace Xamarin.Android.Prepare
+{
+	class GeneratedSourceLinkJsonFile : GeneratedFile
+	{
+		IEnumerable<GitSubmoduleInfo> submodules;
+		string xaCommit;
+
+		public GeneratedSourceLinkJsonFile (IEnumerable<GitSubmoduleInfo> submodules, string xaCommit, string outputPath)
+			: base (outputPath)
+		{
+			if (submodules == null)
+				throw new ArgumentNullException (nameof (submodules));
+			this.submodules = submodules;
+			this.xaCommit   = xaCommit;
+		}
+
+		public override void Generate (Context context)
+		{
+			var json    = new StringBuilder ();
+			json.AppendLine ("{");
+			json.AppendLine ("  \"documents\": {");
+
+			foreach (var submodule in submodules.OrderBy (s => s.Name)) {
+				var localPath   = Path.Combine (BuildPaths.XamarinAndroidSourceRoot, submodule.LocalPath);
+
+				var contentUri  = new UriBuilder (submodule.RepositoryUrl);
+				contentUri.Host = "raw.githubusercontent.com";
+				contentUri.Path += $"/{submodule.CommitHash}";
+
+				json.AppendLine ($"    \"{localPath}/*\": \"{contentUri.Uri}/*\",");
+			}
+			json.AppendLine ($"    \"{BuildPaths.XamarinAndroidSourceRoot}/*\": \"https://raw.githubusercontent.com/xamarin/xamarin-android/{xaCommit}/*\"");
```

Comment:
Constructor doesn't check whether `xaCommit` is valid and given that it is used to form a URL, perhaps it should?